src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala

```diff
@@ -290,12 +330,61 @@ class OrderedRDD[T, K, V] private(rdd: RDD[(K, V)],
     assert(newPartEnd.last == n - 1)
     assert(newPartEnd.zip(newPartEnd.tail).forall { case (i, inext) => i <= inext })
 
-    val newRangeBounds = newPartEnd.init.map(end =>
-      orderedPartitioner.rangeBounds(end))
+    val newRangeBounds = newPartEnd.init.map(orderedPartitioner.rangeBounds)
+
+    new OrderedRDD[PK, K, V](new BlockedRDD(persisted, newPartEnd), OrderedPartitioner(newRangeBounds))
+  }
+}
+
+trait OrderedKey[PK, K] extends Serializable {
+  /**
+    * This method must be monotonic in {@code k}
+    */
+  def project(key: K): PK
+
+  implicit def kOrd: Ordering[K]
+
+  implicit def pkOrd: Ordering[PK]
+
+  implicit def kct: ClassTag[K]
+
+  implicit def pkct: ClassTag[PK]
```

Comment:
I just tried this and it doesn't seem to work? I have this:

```scala
trait OrderedKey[PK, K] extends Serializable { self =>
  /**
    * This method must be monotonic in {@code k}
    */
  def project(key: K): PK

  implicit def kOrd: Ordering[K]

  implicit def pkOrd: Ordering[PK]

  implicit def kct: ClassTag[K]

  implicit def pkct: ClassTag[PK]

  object implicits {
    implicit val kOrd = self.kOrd

    implicit val pkOrd = self.pkOrd

    implicit val kct = self.kct

    implicit val pkct = self.pkct
  }
}
```

and this:

```scala
  def apply[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)(implicit kOk: OrderedKey[PK, K]): OrderedRDD[PK, K, V] = {
    import kOk.implicits._
    import Ordering.Implicits._
    // ...
```

and gradle complains:

```
/Users/dking/projects/hail/src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:39: No ClassTag available for K
    val keys = fastKeys.getOrElse(rdd.map(_._1))
                                         ^
/Users/dking/projects/hail/src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:54: No implicit Ordering defined for PK.
          val r = p.max < pnext.min
                    ^
/Users/dking/projects/hail/src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:62: type mismatch;
 found   : scala.collection.mutable.ArraySeq[PK]
 required: Array[?]
 Note: implicit value pkct is not applicable here because it comes after the application point and it lacks an explicit result type
Error occurred in an application involving default arguments.
      val partitioner = OrderedPartitioner(keyInfo.init.map(pi => pi.max))
                                                           ^
/Users/dking/projects/hail/src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala:79: No implicit Ordering defined for K.
            it.toArray.sortBy(_._1).iterator
                             ^
four errors found
:compileScala FAILED
```

I know the third error is because `PK` doesn't have a `ClassTag`.